# Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Importing Data

# Importing Data set orders.csv

In [5]:
df_ords = pd.read_csv(r'/Users/jacob/Desktop/Instacart Basket Analysis/02 Data/Prepared Data/orders_wrangled.csv', index_col = False)

# Creating a Path

In [3]:
path = r'/Users/jacob/Desktop/Instacart Basket Analysis'

# Importing Data set products.csv

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# Create a Dataframe

In [6]:
df_test = pd.DataFrame()

# Create a mixed type Column

In [7]:
df_test['mix'] = ['a', 'b', 1, True ]

# Show head of the new dataframe

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

# Custom Function that searches for mixed-type columns in the dataframe

In [9]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


# Changing the data type

In [10]:
df_test['mix'] = df_test['mix'].astype('str')

# Finding Missing Values in the products.csv dataset

In [12]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

# Subsetting
# Creating a subset of missing values in the products.csv dataset

In [13]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Knowing the dimensions of the products.csv dataset

In [15]:
df_prods.shape

(49693, 5)

# Check frecuency on df_prods dataframe

In [41]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

# In the DataFrame `df_prods`, when we examine the frequency distribution of prices in the 'en' column, we observe some peculiar characteristics. The minimum price is 1, while the maximum price is 99,999. Additionally, the first, second (median), and third quartiles have values of 4.1, 7.1, and 11.1, respectively. 

# The issue here is that the maximum price of 99,999 appears to be an outlier compared to the other values in the distribution. It doesn't align with the overall distribution of prices, where most prices are clustered between 1 and 11.1. Such an extreme outlier can often indicate data entry errors, anomalies, or an entirely different category of products. Therefore, this extreme maximum value doesn't seem to make logical sense within the context of the other price values in the dataset. Further investigation and data validation may be necessary to determine if this value is accurate or if it should be treated as an outlier.

# Creating a new clean data frame of the products.csv dataset 

In [16]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
df_prods_clean.shape

(49677, 5)

# Another way you can drop all missing values

In [18]:
df_prods.dropna(inplace = True)

# Finding Duplicates in the df_prods_clean dataframe

In [19]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# Addressing Duplicates

In [21]:
# Finding the dimensions of the df_prods_clean
df_prods_clean.shape

(49677, 5)

In [22]:
# Creating a new dataframe of the df_prods_clean without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [23]:
# Finding the dimensions of the new dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [25]:
df_prods_clean_no_dups.shape

(49672, 5)

# Exporting df_prods_clean_no_dups 

In [26]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5

# 2. Check frecuency on df_ords dataframe whether anything about the data looks off or should be investigated further.

In [29]:
# checking frequency of df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# Upon examining our frequency table, specifically focusing on the maximum and minimum values, we've noticed some anomalies in the data for "days_since_prior_order." The minimum value is 0, and the maximum is 3 days. However, upon reviewing the percentile range, it becomes apparent that these recordings are unusual. The 50th percentile is reported as 7, which exceeds the maximum value.
# A similar discrepancy is observed in the case of "number of orders," where the minimum is 1 and the maximum is 1. Nevertheless, the percentile values do not align with these boundaries. For example, the 25th percentile is documented as 5, which exceeds the maximum of 1.

# 3. Check for mixed-type data in your df_ords dataframe.

In [33]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [34]:
# checking for mixed data types 
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

# There are no mixed data type in the df_ords dataframe

# 4. If you find mixed-type data, fix it.

# There are no columns with mixed data type in our df_ords data frame

# 5. Checking for missing values in the df_ords dataframe

In [35]:
# Checking for missing values in df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

# Subsetting

In [36]:
# Creating a subset of missing values in the df_ords dataframe for missing values on the column 'days_since_prior_order'
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [37]:
# Print result of the new data frame df_nan_ords
df_nan_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

# We observe a substantial presence of missing values in the "days_since_prior_order". Upon closer inspection of each column, it becomes evident that these NaN values are associated with instances where the "number of orders" equals 1. Since the total count of user_id equals 206209, which coincides with the count of our missing values, it indicates that all user_id instances where "number of orders" is 1 correspond to missing values in the "days_since_prior_order" column.

# This phenomenon aligns with the expectation that customers placing their initial orders wouldn't have any prior order history, resulting in the absence of data in the "days_since_prior_order" column for these cases.

# 6. Address the missing values using an appropriate method.

# This absence of data aligns with instances of first-time orders, thereby offering insights into our customers and their ordering behavior. Consequently, this column retains its relevance, and there is no need to take any action to eliminate this data.

# 7. Run a check for duplicate values in df_ords dataframe

In [38]:
# Checking for duplicates in df_ords
df_dups_ords = df_ords[df_ords.duplicated()]

In [39]:
# Printing results of the new df_dups_ords
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

# No duplicates were found in the df_ords

# 8. Address the duplicates using an appropriate method

# There´s no need to address any duplicates in this dataframe because there are none.

# Export checked data df_ords

In [40]:
# export df_ords as orders_checked
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))